src/main/java/org/prebid/server/bidder/rubicon/RubiconBidder.java

```diff
@@ -1266,27 +1275,102 @@ private void mergeFirstPartyDataFromUser(ExtUser userExt, ObjectNode result) {
+        final Map<String, List<String>> resultSegments = groupBySegtaxValues(relevantSegments, segtaxValues);
+
+        resultSegments.forEach((segtaxValue, segmentIds) -> {
+            final ArrayNode array = target.putArray(segtaxValue);
+            segmentIds.forEach(array::add);
+        });
+    }
+
+    private static Map<String, List<String>> groupBySegtaxValues(Map<Integer, List<String>> segments,
+                                                                 Set<Integer> segtaxValues) {
+
+        return segments.entrySet().stream()
+                .collect(Collectors.toMap(
+                        entry -> resolveSegmentName(entry.getKey(), segtaxValues),
+                        Map.Entry::getValue,
+                        ListUtil::union));
+    }
+
+    private static String resolveSegmentName(Integer taxonomyId, Set<Integer> segtaxValues) {
+        return segtaxValues.contains(taxonomyId) ? SEGTAX_IAB : SEGTAX_TAX + taxonomyId;
+    }
+
+    private static Map.Entry<Integer, Deque<Segment>> getValidSegments(Data data) {
+        final ObjectNode ext = data.getExt();
+        final JsonNode taxonomyId = ext != null ? ext.get(SEGTAX) : null;
+        if (taxonomyId == null || !taxonomyId.isInt()) {
+            return null;
+        }
+
+        final Deque<Segment> segments = getValidOnlySegments(data.getSegment());
+        return CollectionUtils.isNotEmpty(segments) ? Map.entry(taxonomyId.intValue(), segments) : null;
+    }
+
+    private static Deque<Segment> getValidOnlySegments(List<Segment> segments) {
+        return CollectionUtils.isNotEmpty(segments)
+                ? segments.stream()
+                .filter(segment -> StringUtils.isNotBlank(segment.getId()))
+                .collect(Collectors.toCollection(ArrayDeque::new))
+                : null;
+    }
+
+    private static Map<Integer, List<String>> pickRelevantSegments(final Map<Integer, Deque<Segment>> segments) {
+        final Map<Integer, List<String>> result = new HashMap<>();
+        final List<Integer> segmentsKeys = new ArrayList<>(segments.keySet());
+
+        for (int i = 0; i < MAX_NUMBER_OF_SEGMENTS; ) {
+            final List<Integer> emptySegmentKeys = new ArrayList<>();
+
+            for (Integer segmentKey : segmentsKeys) {
+                final Deque<Segment> currentSegments = segments.get(segmentKey);
+                final Segment lastSegment = currentSegments.pollLast();
+                result.computeIfAbsent(segmentKey, key -> new ArrayList<>()).add(lastSegment.getId());
+
+                if (CollectionUtils.isEmpty(currentSegments)) {
+                    emptySegmentKeys.add(segmentKey);
+                }
+
+                i++;
+            }
+
+            if (!emptySegmentKeys.isEmpty()) {
+                segmentsKeys.removeAll(emptySegmentKeys);
+                if (CollectionUtils.isEmpty(segmentsKeys)) {
+                    break;
+                }
+            }
```

Comment:
Now, it may produce more than `MAX_NUMBER_OF_SEGMENTS` segments :p

Reply:
Just use previous approach, but use `while` instead of `for` with the next condition:
```java
    private static Map<Integer, List<String>> pickRelevantSegments(final Map<Integer, Deque<Segment>> segments) {
        final Map<Integer, List<String>> result = new HashMap<>();
        final List<Integer> segmentsKeys = new ArrayList<>(segments.keySet());

        int i = 0;
        int consumedSegmentsCount = 0;

        while (consumedSegmentsCount < MAX_NUMBER_OF_SEGMENTS && !segmentsKeys.isEmpty()) {
            final int segmentsIndex = i % segmentsKeys.size();
            final Integer segmentKey = segmentsKeys.get(segmentsIndex);
            final Deque<Segment> currentSegments = segments.get(segmentKey);

            final Segment lastSegment = currentSegments.pollLast();
            result.computeIfAbsent(segmentKey, key -> new ArrayList<>()).add(lastSegment.getId());
            consumedSegmentsCount++;

            if (currentSegments.isEmpty()) {
                segmentsKeys.remove(segmentKey);
                i--;
            }
            i++;
        }

        return result;
    }
```